### PR TITLE
[shard-map] "support" device_put in shmap bodies (error if nontrivial)

### PR DESCRIPTION
**Pull request by mattjj**

### CODE REVIEW

**Review by jakevdp — APPROVED** [review]
Nice!

File: jax/experimental/shard_map.py [hunk 1]
```diff
@@ -700,6 +701,15 @@ def _debug_callback_eager_rule(mesh, *args, callback: Callable[..., Any],
   return []
 eager_rules[debugging.debug_callback_p] = _debug_callback_eager_rule
 
+def _device_put_eager_rule(mesh, x, *, src, device):
```
Comment (yashk2810) [inline jax/experimental/shard_map.py:704]: nit: src and device are reversed.

so it should be: `(mesh, x, *, device, src)`

Comment (mattjj) [reply jax/experimental/shard_map.py:704]: Thanks, but they're keyword-only, so the order doesn't matter, right?

Comment (yashk2810) [reply jax/experimental/shard_map.py:704]: Maybe but I feel that maintaining the order is less confusing and it doesn't hurt to do that here :)

**Pull request merged**